lib/livebook/session/data.ex

```diff
@@ -838,6 +860,51 @@ defmodule Livebook.Session.Data do
     |> set!(notebook: Notebook.update_cell(data.notebook, cell.id, &Map.merge(&1, attrs)))
   end
 
+  defp maybe_invalidate_bound_cells({data, _} = data_actions, cell, prev_cell) do
+    invalidates_bound_cells? =
+      case cell do
+        %Cell.Input{} -> cell.value != prev_cell.value or cell.name != prev_cell.name
+        _ -> false
+      end
+
+    if invalidates_bound_cells? do
+      bound_cells = bound_cells_with_section(data, cell.id)
+
+      data_actions
+      |> reduce(bound_cells, fn data_actions, {bound_cell, section} ->
+        dependent = dependent_cells_with_section(data, bound_cell.id)
+        mark_cells_as_stale(data_actions, [{bound_cell, section} | dependent])
+      end)
+    else
+      data_actions
+    end
+  end
+
+  defp maybe_queue_bound_cells({data, _} = data_actions, cell, prev_cell) do
+    queue_bound_cells? =
+      case cell do
+        %Cell.Input{} ->
+          cell.reactive and cell.value != prev_cell.value and Cell.Input.validate(cell) == :ok
```

Comment:
Similar here, we should probably move this logic to Cell.Input. Maybe call it new_reactive_value? or something.